File: 5.最长回文子串/solution.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        ans = [0, 0]
        dp = [[False for _ in range(n)] for _ in range(n)]
        for i in range(n - 1, -1, -1):
            for j in range(i, n):
                if s[i] == s[j] and i + 1 >= j:
                    dp[i][j] = True
                    if j - i > ans[1] - ans[0]:
                        ans = [i, j]
                    continue
                if s[i] == s[j] and dp[i + 1][j - 1] == True:
                    dp[i][j] = True
                    if j - i > ans[1] - ans[0]:
                        ans = [i, j]
                    continue
        return s[ans[0]:ans[1] + 1]

File: 5.最长回文子串/solution.cpp
# include "..\leetcode.h"

class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        // 结果用两个位置存储
        pair<int, int> ans{0, 0};
        // dp[i][j] s[i, j]是否为回文子串
        vector<vector<bool>> dp(n, vector<bool>(n, false));
        for(int i = n - 1; i >= 0; i--) {
            for(int j = i; j < n; j++) {
                // 一个字符or两个字符(aa)
                if(s[i] == s[j] && i + 1 >= j) {
                    dp[i][j] = true;
                    if(j - i > ans.second - ans.first) {
                        ans = {i , j};
                    }
                    continue;
                }
                // s[i](ss)s[j] ss为回文子串并且s[i]==s[j]
                if(s[i] == s[j] && dp[i + 1][j - 1] == true) {
                    dp[i][j] = true;
                    if(j - i > ans.second - ans.first) {
                        ans = {i , j};
                    }
                }
            }
        }
        return s.substr(ans.first, ans.second - ans.first + 1);
    }
};